[BottomDrawer] Initial API review

willlarche opened the pull request:
Attempting to do "as little as possible" and use existing UIKit API.

Much to be borrowed from Bottom Sheet and Dialogs.

willlarche committed:
[BottomDrawer] Initial API design commit.

romoore commented on components/BottomDrawer/src/MDCBottomDrawerPresentationController.h:
```diff
@@ -0,0 +1,38 @@
+    presentedViewFullScreen;
+
+/** Is the presented view's preferredContentSize.height completely visible? */
+@property(nonatomic, assign, readonly, getter=isPresentedViewFullyExtended) BOOL
```
Does this return `YES` or `NO` if the height extends into the safe area?

willlarche replied:
Regardless of how tall the view is, or what the safe area insets are, if it is fully extended and has no more content below, this returns `YES`.

romoore replied:
Ack.

ianegordon replied:
How do we expect these will be used?
They may be useful from a development / debugging viewpoint, but I'm wondering if we would want clients changing behavior based on these values.

willlarche replied:
It's been confirmed that there is behavior (motion) that depends on the content knowing if it's fully presented. I'm exploring if that can just be handled by querying the percent of the percent based transition.

romoore commented on components/BottomDrawer/src/MDCBottomDrawerTransitionController.h:
```diff
@@ -0,0 +1,40 @@
+
+// TODO: (larche) Cite spec article.
+/**
+ A controller for scrubbable scrolling (aka dragging) vertical pan gesture-based mnodal transitions
```
Nit/spelling: s/mnodal/modal

willlarche replied:
Thank you.

romoore reviewed (commented):
Do you have an example file to demonstrate how you think clients will call the API?

Commit:
[Catalog] Pad trailing edge inside. (#2242)

The constraints for MDCNodeListViewController incorrectly made the padding ADD width to the contained (inner) view rather than SUBTRACTING that width.

Commit:
[Catalog] RTL for settings dialog. (#2249)

Flipping the debugging panel layout when the layout orientation is right-to-left.

heinberg commented on components/BottomDrawer/src/MDCBottomDrawerPresentationController.h:
```diff
@@ -0,0 +1,38 @@
+ or full screen.
+
+ Also provides a dark, translucent scrim for easy dismissal and free rounding of corners on the
+ presented view that "heal" if the presented view is full screen.
```
Do we want to make the rounding effect optional (i.e. add a BOOL for whether we want it to round corners? Or is this part of the official spec?

willlarche replied:
It's part of the official spec right now but to be conservative, I'll build in a BOOL in private API that we could always expose in the future.

heinberg commented on components/BottomDrawer/src/MDCBottomDrawerTransitionController.h:
```diff
@@ -0,0 +1,40 @@
+ @param bottomAppBar The bottom app bar that initiates and reacts to the presentation of the
+ drawer.
+ */
+- (nonnull instancetype)initWithBottomAppBar:(nonnull MDCBottomAppBarView *)bottomAppBar
```
It looks like this will always require an MDC app bar. Is it possible that some apps will wan this to launch off a custom view? Is there anything strictly requiring this?

ianegordon replied:
Who will own this view?  I would expect it to be a subview of the presenting controller (MDCBottomAppBarController?).  Will we require the presented view controller to be of a certain type?

willlarche replied:
`MDCBottomAppBarController` will indeed own the view and the drawer. So agreed with @ajsecord and @jgunaratne in a working session.

heinberg commented on components/BottomDrawer/src/MDCBottomDrawerPresentationController.h:
```diff
@@ -0,0 +1,38 @@
+ Also provides a dark, translucent scrim for easy dismissal and free rounding of corners on the
+ presented view that "heal" if the presented view is full screen.
+ */
+@interface MDCBottomDrawerPresentationController: NSObject
```
Sorry for my lack of knowledge - how will the user indicate what UIViewController should be hosted inside the controller?

willlarche replied:
That will be done in the API for `MDCAppBarViewController` which actually puts all these pieces together. That's owned by @jgunaratne .

jverkoey committed:
Update Podfile.lock files across the repo to CocoaPods 1.3.1. (#2257)

jgunaratne committed:
Update bottom app bar for iPhone X (#2234)

Commit:
[Collections] Don't inset sections when there is a safe area inset. (#2258)

ajsecord committed:
Cut release candidate.

ajsecord committed:
Added API diff to CHANGELOG.md.

ajsecord committed:
Hand-modified CHANGELOG.md API diff.

ajsecord committed:
Bumped version number to 37.0.1.

ajsecord committed:
Updated change log with the Swift breaking change in NavigationBar.

ajsecord committed:
Bumped version number to 38.0.0.

jverkoey committed:
[FlexibleHeader] Disable content inset adjustments in the horizontal pager scroll view example's horizontal scroll view. (#2264)

If UIKit insets the horizontal paging scroll view then it will allow vertical scrolling on the horizontal view, which is not desired because vertical scrolling is meant to be managed by each scrollview page.

jverkoey committed:
[FlexibleHeader] Only disable automaticallyAdjustsScrollViewInsets on the parent controller when the device is running an OS older than iOS 11. (#2265)

ianegordon commented on components/BottomDrawer/src/MDCBottomDrawerPresentationController.h:
```diff
@@ -0,0 +1,38 @@
+ Also provides a dark, translucent scrim for easy dismissal and free rounding of corners on the
+ presented view that "heal" if the presented view is full screen.
+ */
+@interface MDCBottomDrawerPresentationController: NSObject
```
I would expect an XYZNameHerePresentationController to be of type UIPresentationController, not a delegate.

Naming is hard.

willlarche replied:
That's a typo. Thank you. It's of type `UIPresentationController` now.

willlarche committed:
[TextFields] Updating kitchen sink example to use trailing and leading views. (#2266)

ajsecord committed:
Adding CHANGELOG.md link for PR description.

jverkoey committed:
[Flexible Header] Rework our enforcement of content insets. (#2263)

* WIP adjustment fix.

* Add back missing height adjustment logic from fhv_adjustInsetsForSafeAreaInScrollView.

* Remove conditional.

* Add comment.

* Line length.

* Address unused variable warning.

ajsecord committed:
Merge branch 'release-candidate'

jgunaratne committed:
[BottomAppBar] Set nav bar frame on layout (#2272)

* Set nav bar frame when layoutSubviews called

* Fix indentation

jgunaratne committed:
[BottomAppBar] Remove unnecessary autoresizing mask (#2273)

* Set nav bar frame when layoutSubviews called

* Fix indentation

* Remove unnecessary autoresize mask

jverkoey committed:
Reran ./scripts/release/manage_pods.py update with CocoaPods 1.3.1. (#2275)

ajsecord committed:
Cut release candidate.

ajsecord committed:
Added API diff to CHANGELOG.md.

ajsecord committed:
Hand-modified CHANGELOG.md API diff.

ajsecord committed:
Bumped version number to 38.0.1.

Commit:
[ActivityIndicator] Migrated to MDFi18n. (#2246)

* [ActivityIndicator] Migrated to MDFi18n.

* Replaced RTL dependency in Podspec

Commit:
Added a script to check that umbrella header are used between components (#2241)

willlarche committed:
[TextFields] Fixes clear button not reseting character count (#2277)

* [TextFields] Adding better UIResponder overrides.

* [TextFields] Sending a notification when the clear button is touched like it would if the clear button were UITextField's.

* [TextFields] More specific notification class checks.

ajsecord committed:
Merge branch 'release-candidate'

Commit:
[BottomBar] Migrated to MDFi18n. (#2245)

* [BottomBar] Migrated to MDFi18n.

* Swapped dependencies from MDCRTL to MDFi18n

Commit:
[NavigationBar] Migrated to MDFi18n (#2280)

* Moved src code to MDFi18n

* Updated Podspec deps

* Migrated examples to MDFi18n

Commit:
[AppBar] Migrated to MDFi18n. (#2243)

* [AppBar] Migrated to MDFi18n.

* Swapped dependencies from MDCRTL to MDFi18n

Commit:
[BottomAppBar] Migrated to MDFi18n. (#2244)

* [BottomAppBar] Migrated to MDFi18n.

* Swapped dependencies from MDCRTL to MDFi18n

Commit:
[CollectionCells] Migrated to MDFi18n. (#2248)

* [CollectionCells] Migrated to MDFi18n.

* Swapped dependencies from MDCRTL to MDFi18n

jverkoey committed:
Update MotionTransitioning dependency to 4.0. (#2238)

* Update MotionTransitioning dependency to 4.0.

* Update Podfile.lock files.

Commit:
[Catalog] De-hardcode the status bar height from a couple of examples. (#2292)

* [Catalog] De-hardcode the status bar height from a couple of examples.

* [Catalog] Fix indentation.

Commit:
[ProgressBar] Migrated to MDFi18n (#2281)

* Migrated code to MDFi18n

* Updated Podspec deps

jverkoey committed:
[MaskedTransition] Use the provided key path consts from MotionAnimator. (#2294)

Commit:
[Tabs] Migrated to MDFi18n (#2282)

Commit:
[TextFields] Migrated to MDFi18n (#2283)

willlarche committed:
[TextFields] Adds cursor color as parameter (#2297)

* [TextFields] Adding cursorColor to example.

* [TextFields] Adding cursor color to tests.

* [TextFields] Completing implementation of cursor color.

Commit:
[ThumbTrack] Migrated to MDFi18n (#2284)

jverkoey committed:
[MaskedTransition] Fix bug where transitions wouldn't complete. (#2293)

This bug was introduced by the upgrade to MotionTransitioning 4.0. 4.0 requires that if a presentation controller implements MDMTransition, that its start method eventually invokes transitionDidEnd.

jgunaratne committed:
[BottomNavigation] Adding bottom navigation component (#2088)

* Adding bottom nav component

* Add bottom navigation item

* Add nav item and cell badge classes

* Add badge color

* Update Bottom Nav Item

* Use KVO

* Update centering for orientation change

* Update badgeValue property

* Add nav bar container view for items

* Check text height when setting frame

* Update badge height calculation

* Add select and unselected color, add icons for demo, RTL support

* Add MDCMath as dependency

* Add demo description to example

* Add selection states for hiding item titles on select

* Change hide state name to default

* Rename view to bar

* Update nullability

* Rename properties for consistency with UITabBar

* Rename container view

* Update selected item property name

* Rename properties for consistency with UITabBar

* Rename classes, formatting

* Replace custom badge layer path with corner radius

* Remove MDCMath as dependency, update formatting

* Update layout to accommodate iPhone X

* Remove unused constant

* Move autoresizing to init

* Update landscape mode layout, add interim selection state

* Remove safeAreaInsetsDidChange

* Separate layout from setSelected

* Update landscape item layout

* Add landscape layout states, update sizeThatFits

* Update some setters, properties and formatting

* Update to dot setter syntax

* Change how image tintColor is modified

* Add i18n and a11y support for VoiceOver

* Update s/copy/strong

* Add a11y selected indicator

* Minor formatting updates

* Minor formatting

* Fix order of operations for remove item views

* Only permit badge color change if iOS 10 or higher.

* Add support for RTL for item adjacent titles

* Spelling correction in comments.

* Update title visibility enum names

* Update title visibility enum names

* Update title visibility names

* Remove 3-5 item requirement and change to recommendation

* nullability updates, distribution mode updates

* Add UIAppearance

* Lower bottom nav height when using adjacent titles

* Changing color/font properties to nonnull.

* Update bottom nav bar main view so it's not an a11y element

* Add shadow to bottom nav

* Make imageview non a11y element.

* Adding tab a11y string

* Remove comment

* Rename distribution enum to alignment, stylistic updates

* Move badgeValue out of setters section

* Add check for iOS 10 when using badge color

* Remove newer availability checks

* Remove unncessary call to viewSafeAreaInsetsDidChange

* Add @available(iOS 11.0, *) check

* Fix badge color iOS version check

* Fix minor typo

* Add shadow layer and elevation dependency to podspec

* Remove (@available(iOS 10.0, *))

willlarche committed:
[TextFields] Major improvements to RTL (#2274)

* [TextFields] Correcting comment.

* [TextFields] Major corrections to RTL by using MDFInternationalization and respecting iOS 8's weird behavior.

* [TextFields] Adding MDFInternationalization to podspec for TextFields.

* [TextFields] Correcting headers.

* [TextFields] Adding a comment.

Commit:
[NavigationBarExamples] Swapped private/RTL with MDFi18n (#2300)

* [NavigationBarExamples] Swapped private/RTL with MDFi18n

* Fix breakage.

* For real this time

* Changed import order.

Commit:
[PageControl] safeArea fixes (#2301)

* PageControl safeArea fixes

* Indentation Pass

jgunaratne committed:
[BottomNavigation] Create Bottom Navigation Swift example (#2288)

* Adding bottom nav component

* Add bottom navigation item

* Add nav item and cell badge classes

* Add badge color

* Update Bottom Nav Item

* Use KVO

* Update centering for orientation change

* Update badgeValue property

* Add nav bar container view for items

* Check text height when setting frame

* Update badge height calculation

* Add select and unselected color, add icons for demo, RTL support

* Add MDCMath as dependency

* Add demo description to example

* Add selection states for hiding item titles on select

* Change hide state name to default

* Rename view to bar

* Update nullability

* Rename properties for consistency with UITabBar

* Rename container view

* Update selected item property name

* Rename properties for consistency with UITabBar

* Rename classes, formatting

* Replace custom badge layer path with corner radius

* Remove MDCMath as dependency, update formatting

* Update layout to accommodate iPhone X

* Remove unused constant

* Move autoresizing to init

* Update landscape mode layout, add interim selection state

* Remove safeAreaInsetsDidChange

* Separate layout from setSelected

* Update landscape item layout

* Add landscape layout states, update sizeThatFits

* Update some setters, properties and formatting

* Update to dot setter syntax

* Change how image tintColor is modified

* Add i18n and a11y support for VoiceOver

* Update s/copy/strong

* Add a11y selected indicator

* Minor formatting updates

* Minor formatting

* Fix order of operations for remove item views

* Only permit badge color change if iOS 10 or higher.

* Add support for RTL for item adjacent titles

* Spelling correction in comments.

* Update title visibility enum names

* Update title visibility enum names

* Update title visibility names

* Remove 3-5 item requirement and change to recommendation

* nullability updates, distribution mode updates

* Add UIAppearance

* Lower bottom nav height when using adjacent titles

* Changing color/font properties to nonnull.

* Update bottom nav bar main view so it's not an a11y element

* Add shadow to bottom nav

* Make imageview non a11y element.

* Adding tab a11y string

* Remove comment

* Rename distribution enum to alignment, stylistic updates

* Move badgeValue out of setters section

* Add check for iOS 10 when using badge color

* Remove newer availability checks

* Remove unncessary call to viewSafeAreaInsetsDidChange

* Add @available(iOS 11.0, *) check

* Add bottom navigation Swift example

* Fix badge color iOS version check

* Fix minor typo

* Add shadow layer and elevation dependency to podspec

* Remove (@available(iOS 10.0, *))

* Stylistic updates

mpetrov committed:
@objc annoaing catalog by convention (#2305)

Commit:
[BotteomNavigation] Migrated to MDFi18n. (#2306)

Commit:
[MaterialRTL] Removed private/RTL (#2307)

* [MaterialRTL] Removed private/RTL

* Cleaned up podspec file

willlarche committed:
[BottomDrawer] Removing init.

willlarche committed:
[BottomDrawer] Spelling correction.

willlarche committed:
[BottomDrawer] Correcting typo.

willlarche committed:
[BottomDrawer] Continuing decoupling of drawer to BAB.

Commit:
[Tabs] Internal fixes to account for the Safe Area insets. (#2261)

* [TabBar] Fixes for Safe Area insets.

* [TabBar] Check bar position.

* [TabBar] Add back intrinsicContentSize for consistency with sizeThatFits.

* [TabBar] RTL fixes

* [TabBar] Call setNeedsUpdateConstraints to match UIKit.

* [TabBar] Remove need for maintaining a height constraint for auto layout.

* [TabBar] Undo sizeThatFits/intrinsicContentSize changes.

* [TabBar] Revert more things.

* [TabBar] Undo more stuff.

* [TabBar] Review

jverkoey committed:
[FlexibleHeader] Update the min/max height before enforcing content insets. (#2312)

This resolves a class of bugs that occur during device orientation changes, where the content offset would not stay a consistent value when rotating to landscape and back.

Commit:
[TabBar] Add safeAreaInsets.bottom to tabBarHeight in MDCTabBarViewController. (#2315)

jgunaratne committed:
[BottomNavigation] Add color themer to bottom navigation bar (#2239)

* Adding bottom nav component

* Add bottom navigation item

* Add nav item and cell badge classes

* Add badge color

* Update Bottom Nav Item

* Use KVO

* Update centering for orientation change

* Update badgeValue property

* Add nav bar container view for items

* Check text height when setting frame

* Update badge height calculation

* Add select and unselected color, add icons for demo, RTL support

* Add MDCMath as dependency

* Add demo description to example

* Add selection states for hiding item titles on select

* Change hide state name to default

* Rename view to bar

* Update nullability

* Rename properties for consistency with UITabBar

* Rename container view

* Update selected item property name

* Rename properties for consistency with UITabBar

* Rename classes, formatting

* Replace custom badge layer path with corner radius

* Remove MDCMath as dependency, update formatting

* Update layout to accommodate iPhone X

* Remove unused constant

* Move autoresizing to init

* Update landscape mode layout, add interim selection state

* Remove safeAreaInsetsDidChange

* Separate layout from setSelected

* Update landscape item layout

* Add landscape layout states, update sizeThatFits

* Update some setters, properties and formatting

* Update to dot setter syntax

* Change how image tintColor is modified

* Add i18n and a11y support for VoiceOver

* Update s/copy/strong

* Add a11y selected indicator

* Minor formatting updates

* Minor formatting

* Fix order of operations for remove item views

* Only permit badge color change if iOS 10 or higher.

* Add support for RTL for item adjacent titles

* Spelling correction in comments.

* Update title visibility enum names

* Update title visibility enum names

* Update title visibility names

* Remove 3-5 item requirement and change to recommendation

* nullability updates, distribution mode updates

* Add UIAppearance

* Lower bottom nav height when using adjacent titles

* Changing color/font properties to nonnull.

* Add bottom navigation color themer

* Update bottom nav bar main view so it's not an a11y element

* Add shadow to bottom nav

* Make imageview non a11y element.

* Adding tab a11y string

* Remove comment

* Rename distribution enum to alignment, stylistic updates

* Move badgeValue out of setters section

* Add check for iOS 10 when using badge color

* Remove newer availability checks

* Remove unncessary call to viewSafeAreaInsetsDidChange

* Add @available(iOS 11.0, *) check

* Fix badge color iOS version check

* Fix minor typo

* Add shadow layer and elevation dependency to podspec

* Remove (@available(iOS 10.0, *))

* Remove code diff

* Update podspec

* Code style updates, improve comments

* Add theming to swift example

willlarche committed:
[TextFields] Fixes bugs on multiline IB support (#2289)

* [TextFields] Improved interface builder support.

* [TextFields] Adding multiline to interface builder example.

* [TextFields] Adding comment.

* [TextFields] Updating IB example.

* [TextFields] Updating storyboard for example.

* [TextFields] Adding handling of expanding on overflow for IB cases.

* [TextFields] Adding better expanding support for IB.

* [TextFields] Fixing bugs related to loading a multiline text field from a storyboard.

* [TextFields] Reorganizing IB examples to add new one.

* [TextFields] Adding useful comment.

jverkoey committed:
[FlexibleHeader] Fix bug where status bar visibility changes could cause the header size to jump. (#2321)

* [FlexibleHeader] Fix bug where status bar visibility changes could cause the header size to jump.

If the status bar visibility changes without the header's awareness - e.g. if the app hides the status bar for any reason - then the logic for maintaining a consistent content offset will not be executed and our flexible header's height will appear to "jump". This only happens if the tracking scroll view is scrollable and the flexible header's max height is > min height.

To resolve this, we need to have some awareness of when the status bar visibility has changed. I'm not aware of a good status bar visibility notification (other than KVO'ing the application's statusBarHidden property), so this change introduces a small amount of state that tracks our last known status bar visibility.

The new logic is intentionally written to only take effect in the case where we know the following to be true:

1. We've observed the status bar visibility at least once.
2. The status bar visibility doesn't match our last known visibility.
3. We are not currently knowingly changing the status bar visibility.

If all of this is true, then when we adjust the content insets for the scroll view we use the inset delta to adjust the content offset by the shifted amount. This counteracts the amount that UIKit will have adjusted our content offset by, resulting in a no-op of content offset changes and our header's height stays constant through the status bar visibility change.

The reason we don't want to use UIKit's default content offset adjusting behavior is because UIKit thinks the content insets should change when the status bar visibility changes. Our flexible header does not allow the content inset to change because doing so would mean our flexible header couldn't expand to its full height. So, when UIKit removes the 20 points from our content inset and shifts our content offset by 20 points, we imemdiately add back the 20 points to our content inset and then (now) reset the content offset back by 20 points again.

* Reset _wasStatusBarHiddenIsValid when moving to a new window.

mpetrov committed:
updating FlexibleHeader readme (#2318)

jverkoey committed:
Initial addition of kokoro and bazel continuous integration support. (#2316)

* Initial addition of kokoro and bazel continuous integration support.

Only adds support for ActivityIndicator and its dependencies. Will add other components to the continuous integration in follow up changes.

* Add AnimationTiming.

* Fix target name.

* Use ... matching to run all tests and build all targets.

* Just run tests.

* Add skeleton tests.

* Fix copyrights and build files.

* Add bzl file.

* Standardize build files.

* Fix typo.

* Naming.

* Remove confusing comment.

* Use Xcode 8.2.

jverkoey committed:
[FlexibleHeader+UIMetrics] Add BUILD file. (#2323)

willlarche commented:
Rebase confused GitHub and CLA bot. Closing and reopening.
